api/services/retention/export_workflow_run.py

```diff
@@ -0,0 +1,314 @@
+"""
+Workflow Run Export Service.
+
+This service handles exporting workflow run data including node executions,
+supporting both archived (from storage) and non-archived (from DB) runs.
+"""
+
+import io
+import json
+import logging
+import zipfile
+from datetime import UTC, datetime
+from typing import Any
+from uuid import uuid4
+
+from sqlalchemy.orm import Session, sessionmaker
+
+from extensions.ext_database import db
+from libs.archive_storage import (
+    ArchiveStorage,
+    ArchiveStorageNotConfiguredError,
+    build_workflow_run_prefix,
+    get_archive_storage,
+)
+from models.workflow import WorkflowRun
+from repositories.api_workflow_run_repository import APIWorkflowRunRepository
+from repositories.factory import DifyAPIRepositoryFactory
+from repositories.sqlalchemy_api_workflow_node_execution_repository import (
+    DifyAPISQLAlchemyWorkflowNodeExecutionRepository,
+)
+from repositories.sqlalchemy_workflow_trigger_log_repository import SQLAlchemyWorkflowTriggerLogRepository
+
+logger = logging.getLogger(__name__)
+
+
+class WorkflowRunExportError(Exception):
+    """Exception raised when export fails."""
+
+    pass
+
+
+class WorkflowRunExportService:
+    """
+    Service for exporting workflow run data.
+
+    Handles both archived and non-archived workflow runs:
+    - Non-archived runs: Data is fetched from the database
+    - Archived runs: Data is fetched from storage
+    """
+
+    EXPORTED_TABLES = [
+        "workflow_app_logs",
+        "workflow_node_executions",
+        "workflow_node_execution_offload",
+        "workflow_pauses",
+        "workflow_pause_reasons",
+        "workflow_trigger_logs",
+    ]
+
+    def __init__(self):
+        self.workflow_run_repo: APIWorkflowRunRepository | None = None
+
+    def export(
+        self,
+        tenant_id: str,
+        run_id: str,
+    ) -> bytes:
+        """
+        Export a workflow run as a ZIP file.
+
+        Args:
+            tenant_id: Tenant ID
+            run_id: Workflow run ID
+
+        Returns:
+            ZIP file bytes containing all workflow run data
+
+        Raises:
+            WorkflowRunExportError: If export fails
+        """
+        session_maker = sessionmaker(bind=db.engine, expire_on_commit=False)
+        repo = self._get_workflow_run_repo()
+
+        with session_maker() as session:
+            run = repo.get_workflow_run_by_id_without_tenant(run_id)
+
+            if not run:
+                raise WorkflowRunExportError(f"Workflow run {run_id} not found")
+
+            if run.tenant_id != tenant_id:
+                raise WorkflowRunExportError(f"Workflow run {run_id} does not belong to tenant {tenant_id}")
+
+            if run.is_archived:
+                return self._export_from_s3(session, run)
+            return self._export_from_db(session, run)
+
+    def export_to_storage(
```

Comment:
we should check if s3 or r2 can configure the lifecycle rule to delete the outdated temporarily uploaded objects.